share/core.go

```diff
@@ -0,0 +1,315 @@
+type PubPoly struct {
+	g       abstract.Group   // Cryptographic group
+	b       abstract.Point   // Base point, nil for standard base
+	commits []abstract.Point // Commitments to coefficients of the secret sharing polynomial
+}
+
+// NewPubPoly creates a new public commitment polynomial
+func NewPubPoly(g abstract.Group, b abstract.Point, commits []abstract.Point) *PubPoly {
+	return &PubPoly{g, b, commits}
+}
+
+// Info returns the base point and the commitments to the polynomial coefficients.
+func (p *PubPoly) Info() (abstract.Point, []abstract.Point) {
+	return p.b, p.commits
+}
+
+// Threshold returns the secret sharing threshold.
+func (p *PubPoly) Threshold() int {
+	return len(p.commits)
+}
+
+// GetCommit returns the secret commitment p(0), i.e., the constant term of the polynomial.
+func (p *PubPoly) GetCommit() abstract.Point {
+	return p.commits[0]
+}
+
+// Eval computes the public share v = p(i).
+func (p *PubPoly) Eval(i int) *PubShare {
+	xi := p.g.Scalar().SetInt64(1 + int64(i)) // x-coordinate of this share
+	v := p.g.Point().Null()
+	for j := p.Threshold() - 1; j >= 0; j-- {
+		v.Mul(v, xi)
+		v.Add(v, p.commits[j])
+	}
+	return &PubShare{i, v}
+}
+
+// Shares creates a list of n public commitment shares p(1),...,p(n).
+func (p *PubPoly) Shares(n int) []*PubShare {
+	shares := make([]*PubShare, n)
+	for i := 0; i < n; i++ {
+		shares[i] = p.Eval(i)
+	}
+	return shares
+}
+
+// Add computes the component-wise sum of the polynomials p and q and returns it
+// as a new polynomial. If the base points p.b and q.b are different then the
+// base point of the resulting PubPoly cannot be computed without knowing the
+// discrete logarithm between p.b and q.b. In this particular case, we are using
+// p.b as a default value which of course does not correspond to the correct
+// base point and thus should not be used in further computations.
+func (p *PubPoly) Add(q *PubPoly) (*PubPoly, error) {
+
+	if p.g != q.g {
+		return nil, errorGroups
+	}
+
+	if p.Threshold() != q.Threshold() {
+		return nil, errorCoeffs
+	}
+
+	t := p.Threshold()
+	commits := make([]abstract.Point, t)
+	for i := 0; i < t; i++ {
+		commits[i] = p.g.Point().Add(p.commits[i], q.commits[i])
+	}
+
+	return &PubPoly{p.g, p.b, commits}, nil
+}
+
+// Equal checks equality of two public commitment polynomials p and q.
+func (p *PubPoly) Equal(q *PubPoly) bool {
+
+	if p.g.String() != q.g.String() {
+		return false
+	}
+
+	for i := 0; i < p.Threshold(); i++ {
+		if !p.commits[i].Equal(q.commits[i]) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// Check a private share against a public commitment polynomial.
+func (p *PubPoly) Check(s *PriShare) bool {
+	pv := p.Eval(s.I)
+	ps := p.g.Point().Mul(p.b, s.V)
+	return pv.V.Equal(ps)
+}
+
+// RecoverCommit reconstructs the secret commitment p(0) using Lagrange interpolation.
+func RecoverCommit(g abstract.Group, shares []*PubShare, t int, n int) (abstract.Point, error) {
```

Comment:
For `RecoverSecret` you are doing Lagrange interpolation over a `[]abstract.Scalar` whereas for `RecoverCommit` you do it over a `[]abstract.Point` and utilise the homomorphic properties of `abstract.Point` addition. Conceptually the techniques are similar but I don't see how it can be generalised enough to make it applicable to both cases except with weird `if-then-else` constructions and using `reflect`. Not sure if it's worth it?